README.md
# Docker Image Cleaner

[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/jupyterhub/docker-image-cleaner/Publish?logo=github)](https://github.com/jupyterhub/docker-image-cleaner/actions)
[![Latest PyPI version](https://img.shields.io/pypi/v/docker-image-cleaner?logo=pypi)](https://pypi.python.org/pypi/docker-image-cleaner)
[![Latest quay.io image tags](https://img.shields.io/github/v/tag/jupyterhub/docker-image-cleaner?include_prereleases&label=quay.io)](https://quay.io/repository/jupyterhub/docker-image-cleaner?tab=tags)

A Python package (`docker-image-cleaner`) and associated Docker image
(`quay.io/jupyterhub/docker-image-cleaner`) to clean up old docker images when a
disk is running low on inodes or space.

The script has initially been developed to help installations of BinderHub clean
up space on nodes as it otherwise can run out of space and stop being able to
build now docker images.

## Why?

Container images are one of the biggest consumers of disk space
and inodes on kubernetes nodes. Kubernetes tries to make sure there is enough
disk space on each node by [garbage
collecting](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#containers-images)
unused container images and containers. Tuning this is important
for [binderhub](https://github.com/jupyterhub/binderhub/) installations,
as many images are built and used only a couple times. However, on
most managed kubernetes installations (like GKE, EKS, etc), we can not
tune these parameters!

This script approximates the specific parts of the kubernetes container image
garbage collection in a configurable way.

## Requirements

1. Only kubernetes nodes using the `docker` runtime are supported.
   `containerd` or `cri-o` container backends are not supported.
2. The script expects to run in a kubernetes `DaemonSet`, with `/var/lib/docker`
   from the node mounted inside the container. This lets the script figure
   out how much disk space docker container images are actually using.
3. The `DaemonSet` should have a `ServiceAccount` attached that has permissions
   to talk to the kubernetes API and cordon / uncordon nodes. This makes sure
   new pods are not scheduled on to the node while image cleaning is happening,
   as it can take a while.

## How does it work?

1. Compute how much space `/var/lib/docker` directory (specified by the
   `DOCKER_IMAGE_CLEANER_PATH_TO_CHECK` environment variable) is taking up.
2. If the disk space used is greater than the garbage collection trigger threshold
   (specified by `DOCKER_IMAGE_CLEANER_THRESHOLD_HIGH`), garbage collection is triggered.
   If not, the script just waits another 5 minutes (set by `DOCKER_IMAGE_CLEANER_INTERVAL_SECONDS`).
3. If garbage collection is triggered, the kubernetes node is first cordoned
   to prevent any new pods from being scheduled on it for the duration of the
   garbage collection.
4. Stopped containers are removed via `docker container prune`.
5. Dangling images are removed via `docker image prune`
6. If no dangling images are found to prune, _all_ images are pruned (`docker image prune -a`)
7. After the garbage collection is done, the kubernetes node is also uncordoned.
8. When done, we wait another 5 minutes (set by `DOCKER_IMAGE_CLEANER_INTERVAL_SECONDS`), and repeat
   the whole process.

## Configuration options

Currently, environment variables are used to set configuration for now.

| Env variable                            | Description                                                                                                                 | Default           |
| --------------------------------------- | --------------------------------------------------------------------------------------------------------------------------- | ----------------- |
| `DOCKER_IMAGE_CLEANER_NODE_NAME`        | The k8s node where the docker image cleaner is running, so it can be cordoned via the k8s api                               |                   |
| `DOCKER_IMAGE_CLEANER_PATH_TO_CHECK`    | Path to `/var/lib/docker` directory used by the docker daemon                                                               | `/var/lib/docker` |
| `DOCKER_IMAGE_CLEANER_INTERVAL_SECONDS` | Amount of time (in seconds) to wait between checking if GC needs to be triggered                                            | `300`             |
| `DOCKER_IMAGE_CLEANER_DELAY_SECONDS`    | Amount of time (in seconds) to wait between deleting container images, so we don't DOS the docker API                       | `1`               |
| `DOCKER_IMAGE_CLEANER_THRESHOLD_TYPE`   | Determine if GC should be triggered based on relative or absolute disk usage                                                | `relative`        |
| `DOCKER_IMAGE_CLEANER_THRESHOLD_HIGH`   | % or absolute disk space available (based on `DOCKER_IMAGE_CLEANER_THRESHOLD_TYPE`) when we start deleting container images | `80`              |
| `DOCKER_IMAGE_CLEANER_TIMEOUT_SECONDS`  | Request timeout (in seconds) for docker API requests. Pruning images often takes minutes. Default: 300 (5 minutes)          |


CHANGELOG.md
# Changelog

## 1.0

### 1.0.0-beta.2 - 2022-06-11

([full changelog](https://github.com/jupyterhub/docker-image-cleaner/compare/1.0.0-beta.1...1.0.0-beta.2))

#### Breaking changes

All environment variables configuring the docker-image-cleaner has been renamed
but have the same functionality.

| Before                    | After                                   |
| ------------------------- | --------------------------------------- |
| `NODE_NAME`               | `DOCKER_IMAGE_CLEANER_NODE_NAME`        |
| `PATH_TO_CHECK`           | `DOCKER_IMAGE_CLEANER_PATH_TO_CHECK`    |
| `IMAGE_GC_INTERVAL`       | `DOCKER_IMAGE_CLEANER_INTERVAL_SECONDS` |
| `IMAGE_GC_DELAY`          | `DOCKER_IMAGE_CLEANER_DELAY_SECONDS`    |
| `IMAGE_GC_THRESHOLD_TYPE` | `DOCKER_IMAGE_CLEANER_THRESHOLD_TYPE`   |
| `IMAGE_GC_THRESHOLD_HIGH` | `DOCKER_IMAGE_CLEANER_THRESHOLD_HIGH`   |
| `IMAGE_GC_THRESHOLD_LOW`  | `DOCKER_IMAGE_CLEANER_THRESHOLD_LOW`    |

#### Bugs fixed

- prune all images if no dangling images left [#51](https://github.com/jupyterhub/docker-image-cleaner/pull/51) ([@minrk](https://github.com/minrk))

#### Maintenance and upkeep improvements

- breaking, maint: prefix env vars with project name [#55](https://github.com/jupyterhub/docker-image-cleaner/pull/55) ([@consideRatio](https://github.com/consideRatio))
- pre-commit: use isort over reorder-python-imports [#53](https://github.com/jupyterhub/docker-image-cleaner/pull/53) ([@consideRatio](https://github.com/consideRatio))
- Bump actions/setup-python from 3 to 4 [#50](https://github.com/jupyterhub/docker-image-cleaner/pull/50) ([@dependabot](https://github.com/dependabot))

#### Other merged PRs

- Refreeze Dockerfile's requirements.txt [#58](https://github.com/jupyterhub/docker-image-cleaner/pull/58) ([@jupyterhub-bot](https://github.com/jupyterhub-bot))
- ci: fix detail in refreeze requirements.txt workflow [#57](https://github.com/jupyterhub/docker-image-cleaner/pull/57) ([@consideRatio](https://github.com/consideRatio))
- image: use frozen requirements.txt and automate updates of it [#56](https://github.com/jupyterhub/docker-image-cleaner/pull/56) ([@consideRatio](https://github.com/consideRatio))
- breaking, maint: require py38+ [#54](https://github.com/jupyterhub/docker-image-cleaner/pull/54) ([@consideRatio](https://github.com/consideRatio))

#### Contributors to this release

([GitHub contributors page for this release](https://github.com/jupyterhub/docker-image-cleaner/graphs/contributors?from=2022-05-25&to=2022-06-11&type=c))

[@consideRatio](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3AconsideRatio+updated%3A2022-05-25..2022-06-11&type=Issues) | [@dependabot](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3Adependabot+updated%3A2022-05-25..2022-06-11&type=Issues) | [@jupyterhub-bot](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3Ajupyterhub-bot+updated%3A2022-05-25..2022-06-11&type=Issues) | [@minrk](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3Aminrk+updated%3A2022-05-25..2022-06-11&type=Issues) | [@welcome](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3Awelcome+updated%3A2022-05-25..2022-06-11&type=Issues) | [@yuvipanda](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3Ayuvipanda+updated%3A2022-05-25..2022-06-11&type=Issues)

### 1.0.0-beta.1 - 2022-05-25

#### Enhancements made

- use docker system prune to clean images [#37](https://github.com/jupyterhub/docker-image-cleaner/pull/37) ([@minrk](https://github.com/minrk), [@consideRatio](https://github.com/consideRatio), [@betatim](https://github.com/betatim))

#### Maintenance and upkeep improvements

- Bump kubernetes from 23.3.0 to 23.6.0 [#47](https://github.com/jupyterhub/docker-image-cleaner/pull/47) ([@dependabot](https://github.com/dependabot), [@consideRatio](https://github.com/consideRatio))
- Bump kubernetes from 21.7.0 to 23.3.0 [#32](https://github.com/jupyterhub/docker-image-cleaner/pull/32) ([@dependabot](https://github.com/dependabot), [@minrk](https://github.com/minrk))
- Bump kubernetes from 20.13.0 to 21.7.0 [#23](https://github.com/jupyterhub/docker-image-cleaner/pull/23) ([@dependabot](https://github.com/dependabot), [@consideRatio](https://github.com/consideRatio))
- Bump kubernetes from 19.15.0 to 20.13.0 [#21](https://github.com/jupyterhub/docker-image-cleaner/pull/21) ([@dependabot](https://github.com/dependabot), [@consideRatio](https://github.com/consideRatio))
- Bump kubernetes from 18.20.0 to 19.15.0 [#17](https://github.com/jupyterhub/docker-image-cleaner/pull/17) ([@dependabot](https://github.com/dependabot), [@consideRatio](https://github.com/consideRatio))

#### Contributors to this release

([GitHub contributors page for this release](https://github.com/jupyterhub/docker-image-cleaner/graphs/contributors?from=2021-10-10&to=2022-05-25&type=c))

[@betatim](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3Abetatim+updated%3A2021-10-10..2022-05-25&type=Issues) | [@consideRatio](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3AconsideRatio+updated%3A2021-10-10..2022-05-25&type=Issues) | [@minrk](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3Aminrk+updated%3A2021-10-10..2022-05-25&type=Issues) | [@yuvipanda](https://github.com/search?q=repo%3Ajupyterhub%2Fdocker-image-cleaner+involves%3Ayuvipanda+updated%3A2021-10-10..2022-05-25&type=Issues)

### 1.0.0-alpha.2 - 2021-10-10

Initial fixes to make it function properly in jupyterhub/binderhub and jupyterhub/mybinder.org-deploy.

### 1.0.0-alpha.1 - 2021-09-21

This script has lived by itself as part of jupyterhub/binderhub until now, but
going onwards it is to live in this standalone GitHub repository and be
published as a PyPI package and Docker image on quay.io.


docker_image_cleaner/cleaner.py
"""
Clean docker images

Calls docker prune when the disk starts getting full.

This serves as a substitute for Kubernetes ImageGC
which has thresholds that are not sufficiently configurable on GKE
at this time.
"""
import logging
import os
import time
from contextlib import contextmanager, nullcontext
from functools import partial

import docker
import requests

logging.basicConfig(format="%(asctime)s %(message)s", level=logging.INFO)


annotation_key = "hub.jupyter.org/image-cleaner-cordoned"


def get_absolute_size(path):
    """
    Directory size in gigabytes
    """
    total = 0
    for dirpath, dirnames, filenames in os.walk(path):
        for fname in filenames:
            f = os.path.join(dirpath, fname)
            # some files are links, skip them as they don't use
            # up additional space
            if os.path.isfile(f):
                total += os.path.getsize(f)

    return total / (2**30)


def get_used_percent(path):
    """
    Return disk usage as a percentage

    (100 is full, 0 is empty)

    Calculated by blocks or inodes,
    which ever reports as the most full.
    """
    stat = os.statvfs(path)
    inodes_avail = stat.f_favail / stat.f_files
    blocks_avail = stat.f_bavail / stat.f_blocks
    return 100 * (1 - min(blocks_avail, inodes_avail))


def cordon(kube, node):
    """cordon a kubernetes node"""
    logging.info(f"Cordoning node {node}")
    kube.patch_node(
        node,
        {
            # record that we are the one responsible for cordoning
            "metadata": {
                "annotations": {
                    annotation_key: "true",
                },
            },
            "spec": {
                "unschedulable": True,
            },
        },
    )


def uncordon(kube, node):
    """uncordon a kubernetes node"""
    logging.info(f"Uncordoning node {node}")
    kube.patch_node(
        node,
        {
            # clear annotation since we're no longer the reason for cordoning,
            # if the node ever does become cordoned
            "metadata": {
                "annotations": {
                    annotation_key: None,
                },
            },
            "spec": {
                "unschedulable": False,
            },
        },
    )


@contextmanager
def cordoned(kube, node):
    """Context manager for cordoning a node"""
    try:
        cordon(kube, node)
        yield
    finally:
        uncordon(kube, node)


def main():

    node = os.getenv("DOCKER_IMAGE_CLEANER_NODE_NAME")
    if node:
        import kubernetes.client
        import kubernetes.config

        try:
            kubernetes.config.load_incluster_config()
        except Exception:
            kubernetes.config.load_kube_config()
        kube = kubernetes.client.CoreV1Api()
        # verify that we can talk to the node
        node_info = kube.read_node(node)
        # recover from possible crash!
        if node_info.spec.unschedulable and node_info.metadata.annotations.get(
            annotation_key
        ):
            logging.warning(
                f"Node {node} still cordoned, possibly leftover from earlier crash of image-cleaner"
            )
            uncordon(kube, node)

        cordon_context = partial(cordoned, kube, node)
    else:
        cordon_context = nullcontext

    path_to_check = os.getenv("DOCKER_IMAGE_CLEANER_PATH_TO_CHECK", "/var/lib/docker")
    interval_seconds = float(os.getenv("DOCKER_IMAGE_CLEANER_INTERVAL_SECONDS", "300"))
    delay_seconds = float(os.getenv("DOCKER_IMAGE_CLEANER_DELAY_SECONDS", "1"))
    threshold_type = os.getenv("DOCKER_IMAGE_CLEANER_THRESHOLD_TYPE", "relative")
    threshold_high = float(os.getenv("DOCKER_IMAGE_CLEANER_THRESHOLD_HIGH", "80"))
    timeout_seconds = int(os.getenv("DOCKER_IMAGE_CLEANER_TIMEOUT_SECONDS", "300"))

    docker_client = docker.from_env(version="auto", timeout=timeout_seconds)

    # with the threshold type set to relative the thresholds are interpreted
    # as a percentage of how full the partition is. In absolute mode the
    # thresholds are interpreted as size in bytes. By default you should use
    # "relative" mode. Use "absolute" mode when you are using DIND and your
    # nodes only have one partition.
    if threshold_type == "relative":
        get_used = get_used_percent
        used_msg = "{used:.1f}% used"
        threshold_s = f"{threshold_high}% inodes or blocks"
    else:
        get_used = get_absolute_size
        used_msg = "{used:.2f}GB used"
        threshold_s = f"{threshold_high // (2**30):.0f}GB"

        if threshold_high <= 2**30:
            raise ValueError(
                f"Absolute GC threshold should be at least 1GB, got {threshold_high}B"
            )
        # units in GB
        threshold_high = threshold_high / (2**30)

    logging.info(f"Pruning docker images when {path_to_check} has {threshold_s} used")

    while True:
        used = get_used(path_to_check)
        logging.info(used_msg.format(used=used))
        if used < threshold_high:
            # Do nothing! We have enough space
            time.sleep(interval_seconds)
            continue

        images = docker_client.images.list(all=True)
        if not images:
            logging.info("No images to delete")
            time.sleep(interval_seconds)
            continue
        else:
            logging.info(f"{len(images)} images available to prune")

        # Ensure the node is cordoned while we prune
        with cordon_context():
            for kind in ("containers", "images"):
                key = f"{kind.title()}Deleted"
                tic = time.perf_counter()
                try:
                    # docker_client.containers.prune: https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.ContainerCollection.prune
                    # docker_client.images.prune: https://docker-py.readthedocs.io/en/stable/images.html#docker.models.images.ImageCollection.prune
                    pruned = getattr(docker_client, kind).prune()
                except requests.exceptions.ReadTimeout:
                    logging.warning(f"Timeout pruning {kind}")
                    # Delay longer after a timeout, which indicates that Docker is overworked
                    time.sleep(max(delay_seconds, 30))
                    continue

                toc = time.perf_counter()
                deleted = pruned[key]

                if not deleted and kind == "images":
                    # No dangling images to delete, still space to free.
                    logging.info("No dangling images to prune, pruning _all_ images")
                    tic = time.perf_counter()
                    try:
                        # prune again, this time with `dangling=False` filter,
                        # which deletes _all_ images instead of just dangling ones
                        pruned = docker_client.images.prune(filters={"dangling": False})
                    except requests.exceptions.ReadTimeout:
                        logging.warning("Timeout pruning all images")
                        # Delay longer after a timeout, which indicates that Docker is overworked
                        time.sleep(max(delay_seconds, 30))
                        continue
                    toc = time.perf_counter()
                    deleted = pruned[key]

                if deleted is None:
                    # returns None instead of empty list when nothing to delete
                    n_deleted = 0
                else:
                    n_deleted = len(deleted)
                duration = toc - tic
                bytes_deleted = pruned["SpaceReclaimed"]
                gb = bytes_deleted / (2**30)
                logging.info(
                    f"Deleted {n_deleted} {kind}, freed {gb:.2f}GB in {duration:.0f} seconds."
                )

        time.sleep(interval_seconds)


if __name__ == "__main__":
    main()
